### PR TITLE
Updated QirRuntime/README.md, build/test instructions

### PR DESCRIPTION
**Pull request by kuzminrobin**

### CODE REVIEW

File: src/QirRuntime/README.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # The Native QIR Runtime
 
-This folder contains QIR runtime project, which includes implementation of the
+This folder contains the Quantum Intermediate Representation (QIR) Runtime project. QIR is a subset of the [LLVM](https://llvm.org/) Intermediate Representation.
+QIR Runtime includes implementation of the
  [QIR specification](https://github.com/microsoft/qsharp-language/tree/main/Specifications/QIR) and the bridge to
  compile QIR to be run against the native full state simulator.
```
Comment (bettinaheim) [inline src/QirRuntime/README.md:6]:
```suggestion
 run QIR against the native full state simulator.
```

File: src/QirRuntime/README.md [hunk 2]
```diff
@@ -11,24 +12,17 @@ This folder contains QIR runtime project, which includes implementation of the
 
 ## Build
 
+### Prerequisites
 The QirRuntime project is using CMake (3.17) + Ninja(1.10.0) + Clang++(11.0.0). Other versions of the tools might work
- but haven't been tested. Only x64 architecture is supported.
+ but haven't been tested. Only x64 architecture is supported.  
+For runing the PowerShell scripts below use the powershell core (`pwsh`),
```
Comment (bettinaheim) [inline src/QirRuntime/README.md:18]: Is this necessary? I believe with a minor tweak to the condition in prerequisites.ps1 the scripts run just fine on the out of the box installation (they did about three weeks ago in any case).
```suggestion
For running the PowerShell scripts below use the powershell core (`pwsh`),
```

File: src/QirRuntime/build.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
+# DEPRECATED !!!
```
Comment (bettinaheim) [inline src/QirRuntime/build.py:4]: Is this script no longer meant to be used? If so, what is the reason to not just remove it?

File: src/QirRuntime/generateqir.py [hunk 4]
```diff
@@ -1,6 +1,8 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
+# DEPRECATED !!!
```
Comment (bettinaheim) [inline src/QirRuntime/generateqir.py:4]: Same question here.

File: src/QirRuntime/test.py [hunk 5]
```diff
@@ -1,6 +1,8 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
+# DEPRECATED !!!
```
Comment (bettinaheim) [inline src/QirRuntime/test.py:4]: ... and here. :)

**Review by bettinaheim — COMMENTED** [review]
I haven't tried to follow the instructions just yet, but will do so a bit later. For now just a few comments reading through it.

File: src/QirRuntime/README.md [hunk 6]
```diff
@@ -1,6 +1,7 @@
 # The Native QIR Runtime
 
-This folder contains QIR runtime project, which includes implementation of the
+This folder contains the Quantum Intermediate Representation (QIR) Runtime project. QIR is a subset of the [LLVM](https://llvm.org/) Intermediate Representation.
```
Comment (cgranade) [inline src/QirRuntime/README.md:3]: You've used the definite article "the" to refer to LLVM IR, but not to QIR; I'd suggest either adding a "the" before "QIR" or dropping the "the" before "LLVM."

File: src/QirRuntime/README.md [hunk 7]
```diff
@@ -1,6 +1,7 @@
...
+QIR Runtime includes implementation of the
```
Comment (cgranade) [inline src/QirRuntime/README.md:4]:
```suggestion
The QIR runtime includes an implementation of the
```

File: src/QirRuntime/README.md [hunk 8]
```diff
@@ -11,24 +12,17 @@ This folder contains QIR runtime project, which includes implementation of the
 
 ## Build
 
+### Prerequisites
```
Comment (cgranade) [inline src/QirRuntime/README.md:15]: Blank lines are recommended after Markdown headers.
```suggestion
### Prerequisites

```

File: src/QirRuntime/README.md (hunk 2 above)
Comment (cgranade) [inline src/QirRuntime/README.md:18]: Note that from version 7.0, "PowerShell Core" is just "PowerShell." (Similar to that .NET 5 is just ".NET" even though it's the next version of .NET Core.)

```suggestion
For runing the PowerShell scripts below use PowerShell Core or PowerShell (`pwsh`),
```

File: src/QirRuntime/README.md [hunk 9]
```diff
@@ -11,24 +12,17 @@ This folder contains QIR runtime project, which includes implementation of the
...
 The QirRuntime project is using CMake (3.17) + Ninja(1.10.0) + Clang++(11.0.0). Other versions of the tools might work
- but haven't been tested. Only x64 architecture is supported.
+ but haven't been tested. Only x64 architecture is supported.  
+For runing the PowerShell scripts below use the powershell core (`pwsh`),
+e.g. [this one](https://github.com/PowerShell/PowerShell/releases/download/v7.1.2/PowerShell-7.1.2-win-x64.msi), not the inbox PowerShell.
```
Comment (cgranade) [inline src/QirRuntime/README.md:19]: This links directly to a specific version of the Windows 64-bit installer; I'd suggest linking to the PowerShell installation page, both so that we don't need to update the link when new versions are released, and so that users on other operating systems aren't directed to Windows-only links. E.g.: https://docs.microsoft.com/powershell/scripting/install/installing-powershell.

File: src/QirRuntime/README.md [hunk 10]
```diff
@@ -11,24 +12,17 @@ This folder contains QIR runtime project, which includes implementation of the
...
 
-You can use CMake directly. For example, to produce a release build:
+To install prerequisite tools for building the QIR runtime, you can set the `ENABLE_QIRRUNTIME` environment variable to the string `"true"` 
+and run [`prerequisites.ps1`](prerequisites.ps1), or manually install pre-reqs with the steps listed below
+(Windows steps rely on [Chocolatey](https://chocolatey.org/) for installation).
```
Comment (cgranade) [inline src/QirRuntime/README.md:23]:
```suggestion
and run [`prerequisites.ps1`](prerequisites.ps1), or manually install pre-reqs with the steps listed below.
Note that on Windows, this script relies on the [Chocolatey package manager](https://chocolatey.org/), while on macOS, `prerequisites.ps1` relies on the `brew` package manager.
```

A link to homebrew may also be helpful here.

File: src/QirRuntime/README.md [hunk 11]
```diff
@@ -55,35 +49,53 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
 
 See [https://code.visualstudio.com/docs/remote/wsl] on how to use VS Code with WSL.
 
-## Test
+#### Other Prerequisites
```
Comment (cgranade) [inline src/QirRuntime/README.md:52]:
```suggestion
#### Other Prerequisites

```

File: src/QirRuntime/README.md [hunk 12]
```diff
@@ -55,35 +49,53 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
...
 
-Some of the tests depend on Microsoft.Quantum.Simulator.Runtime library. To run them make sure to build Native simulator
- from this repository or provide your own version of the library in a folder the OS would search during dynamic library
- lookup.
 
-Some of the tests use generated QIR (*.ll) files as build input. Currently the files are checked-in as part of the project
- but in the future they will be replaced by automatic generation during build. To regenerate the files, run generateqir.py
- or build/test scripts without specifying `noqirgen`. To use the checked-in files without regenerating them, run build/test
- scripts with `noqirgen` argument.
+### Build Commands
```
Comment (cgranade) [inline src/QirRuntime/README.md:58]:
```suggestion
### Build Commands

```

File: src/QirRuntime/README.md [hunk 13]
```diff
@@ -55,35 +49,53 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
...
+```
+  
+The script will create the `build/{Debug|Release}` folder and place the build artifacts in it. The configuration `Debug|Release`
+is specified with the `BUILD_CONFIGURATION` environment variable.
+If the variable is not set then the default is specified in [`set-env.ps1`](../../build/set-env.ps1).  
+We strongly recommend doing local builds using this [`build-qir-runtime.ps1`](build-qir-runtime.ps1) build script 
+because it is used by the continuous integration (CI) infrastructure, and this script also runs clang-tidy if it is installed.
 
-### Running tests with test.py
+Or you can use CMake directly (not recommended). For example in Windows you can enter the following commands in the command line:
```
Comment (cgranade) [inline src/QirRuntime/README.md:70]:
```suggestion
Though it is not recommended, you can also use CMake directly. For example, in Windows you can enter the following commands in the command line:
```

File: src/QirRuntime/README.md [hunk 14]
```diff
@@ -55,35 +49,53 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
...
+```batch
+@rem Navigate to QirRuntime folder.
```
Comment (cgranade) [inline src/QirRuntime/README.md:72]: This looks like it uses cmd notation, but cmd is no longer the default shell on Windows (PowerShell was made the default several years ago).

File: src/QirRuntime/README.md [hunk 15]
```diff
@@ -55,35 +49,53 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
...
-All native tests, including QIR, use catch2 and are fully integrated with CTest. Navigate into
- `build/[Windows|Linux]/[Debug|Release]` folder and run `ctest`. No configuration options required. The results will be
- logged into the corresponding `build/[Windows|Linux]/[Debug|Release]/<target_path>/<test_binary_name>_results.xml` file.
- Tests from the "[skip]" category won't be run.
+@rem Return from the `build` directory:
+popd
+```
+## Tests
+
+### Running All Tests
```
Comment (cgranade) [inline src/QirRuntime/README.md:91]:
```suggestion
### Running All Tests

```

File: src/QirRuntime/README.md [hunk 16]
```diff
@@ -55,35 +49,53 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
...
 
-### Running test binaries individually
+```batch
```
Comment (cgranade) [inline src/QirRuntime/README.md:93]: Similarly here, the `@rem` notation is specific to cmd.exe.

File: src/QirRuntime/build.py [hunk 17]
```diff
@@ -1,6 +1,8 @@
...
+
```
Comment (cgranade) [inline src/QirRuntime/build.py:5]: I'd suggest replacing this with a Python-style docstring:

```suggestion
"""
WARNING: Note that this script has been deprecated.
          Please use the `build-qir-runtime.ps1` script instead.
          See README.md for more details.
"""
```

File: src/QirRuntime/generateqir.py (hunk 4 above)
Comment (cgranade) [inline src/QirRuntime/generateqir.py:4]: Similarly here.

File: src/QirRuntime/test.py (hunk 5 above)
Comment (cgranade) [inline src/QirRuntime/test.py:4]: Similarly here.

File: src/QirRuntime/README.md (hunk 2 above)
Comment (kuzminrobin) [reply src/QirRuntime/README.md:18]: After Stefan significantly refactored our build scripts 1 - 2 weeks ago, for me the .ps1 were failing without `pwsh`.

File: src/QirRuntime/README.md (hunk 14 above)
Comment (kuzminrobin) [reply src/QirRuntime/README.md:72]: Yes... I'm out-of-date ;-)
Don't you have on top of your head, what I should use instead of  ```batch
for PowerShell syntax highlighting? Locally I try `ps`, `ps1`, `PowerShell` but my MarkDown plugin for Chrome does not highlight the syntax if I use any of those.

Comment (kuzminrobin) [reply src/QirRuntime/README.md:72]: Oh. looks like it is shown [here](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/1511).

File: src/QirRuntime/README.md (hunk 10 above)
Comment (kuzminrobin) [reply src/QirRuntime/README.md:23]: > A link to homebrew may also be helpful here.

I assume it is https://brew.sh

File: src/QirRuntime/README.md [hunk 18]
```diff
@@ -55,35 +49,53 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
...
+pushd build
 
-The script will trigger an incremental build unless `nobuild` options is specified. Tests from the "[skip]" category
- won't be run.
+@rem Prepare for the build:
+cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug ..
 
-### Running tests with CTest
+@rem Build (you will likely need to issue this command several times, see issue #589):
+cmake --build .
```
Comment (swernli) [inline src/QirRuntime/README.md:89]: I still don't understand this. Why does this need to be run multiple times? I haven't observed that behavior, though I'm trying now on a fresh repo to see if there is something I'm missing.

File: README.md [hunk 19]
```diff
@@ -57,6 +57,37 @@ To build on other platforms:
 1. Install the pre-reqs:
     * Install [CMake](https://cmake.org/install/)
     * Install [.NET Core 3 SDK](https://dotnet.microsoft.com/download)
+    * On [WSL](https://docs.microsoft.com/en-us/windows/wsl/)/Linux:
+      * Install `g++` (e.g. in Ubuntu 20.04 `sudo apt-get install g++`).
+      * The build does not accept `dotnet-*-5.0`/`aspnetcore-*-5.0` packages, install `dotnet-*-3.1`/`aspnetcore-*-3.1`
+        (`sudo apt-get install aspnetcore-runtime-3.1 dotnet-sdk-3.1`). The possible result can be:
```
Comment (swernli) [inline README.md:63]: is the `aspnetcore-runtime-3.1` really necessary?

Comment (kuzminrobin) [reply README.md:63]: You're right. 
> sudo apt-get install **dotnet-sdk-3.1**

is enough.

File: src/QirRuntime/README.md (hunk 18 above)
Comment (kuzminrobin) [reply src/QirRuntime/README.md:89]: I saw similar behavior during multithreaded builds ( `make -j ..` ). Different targets are built by different threads. And sometimes the later targets (that depend on the results of the earlier target builds) start earlier than the earlier target build  fully completes. It felt like the earlier target step was completed (from the point of view of the build utility), but the file output of the earlier stage was cached for a few seconds (e.g. the object file was not written to the disk completely), but the linking stage was already kicked off, and the linker was failing to open the object file, and the linking was failing.
At that moment the cached file was flushed onto the disk and the repeated build attempt was succeeding.

Comment (swernli) [reply src/QirRuntime/README.md:89]: I haven't observed this behavior with CMake, which does not perform parallel builds by default. In fact, CMake will build in dependency order based on detected dependencies across different targets, with the order of the folders in the top level CMakeLists.txt being the fallback. This ensures that CMake will only need to be run once. Have you gotten this error in your local builds?

I did just try it again, and after running bootstrap.ps1 and copying the relevant .ll files to the correct locations, I was able to manually create `src\QirRuntime\build\Debug` and then run these two commands to successfully build the whole QIR runtime and tests:
```
cmake -G Ninja ../..
cmake --build . --target install
```

File: src/QirRuntime/README.md [hunk 20]
```diff
@@ -55,35 +51,57 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
...
-To execute all tests locally run `test.py` from the project's root folder:
+Though it is not recommended, you can also use CMake directly.
+For example, in Windows you can enter the following commands in the command line:
+```powershell
+# Navigate to QirRuntime folder.
 
-- (Windows) `python test.py [nobuild] [debug/release]`
-- (Linux) `python3 test.py [nobuild] [debug/release]`
+# Creatge the `build` directory:
+mkdir build
```
Comment (swernli) [inline src/QirRuntime/README.md:80]: to match what the script does, the directory should be `build/Debug`, then navigate into that folder.

File: src/QirRuntime/README.md [hunk 21]
```diff
@@ -55,35 +51,57 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
...
 
-The script will trigger an incremental build unless `nobuild` options is specified. Tests from the "[skip]" category
- won't be run.
+# Enter the `build` directory:
+pushd build
 
-### Running tests with CTest
+# Prepare for the build:
+cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug ..
```
Comment (swernli) [inline src/QirRuntime/README.md:86]: This needs to be `cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug ../..` since the pwd should be `src\QirRuntime\build\Debug`

File: src/QirRuntime/README.md (hunk 18 above)
Comment (swernli) [reply src/QirRuntime/README.md:89]: Also worth noting, the `--target install` ensures that all the binary output is collected in the `bin` folder, which the test script expects.

File: src/QirRuntime/README.md (hunk 20 above)
Comment (kuzminrobin) [reply src/QirRuntime/README.md]: I removed those steps, they have some more issues.

File: src/QirRuntime/README.md (hunk 21 above)
Comment (kuzminrobin) [reply src/QirRuntime/README.md]: I removed those steps, they have some more issues.

File: src/QirRuntime/README.md (hunk 18 above)
Comment (kuzminrobin) [reply src/QirRuntime/README.md]: I just removed those steps.

Comment (kuzminrobin) [reply src/QirRuntime/README.md]: I double checked out of curiosity. Looks like `--target install` was the missing part.

**Comment by kuzminrobin** [comment]
All the feedback has been addressed. Please review.

File: README.md [hunk 22]
```diff
@@ -57,6 +57,37 @@ To build on other platforms:
...
+ii  dotnet-apphost-pack-3.1   3.1.13-1     amd64        Microsoft.NETCore.App.Host 3.1.13
+ii  dotnet-host               5.0.4-1      amd64        Microsoft .NET Host - 5.0.4
+ii  dotnet-hostfxr-3.1        3.1.13-1     amd64        Microsoft .NET Core Host FX Resolver - 3.1.13 3.1.13
+un  dotnet-nightly            <none>       <none>       (no description available)
+ii  dotnet-runtime-3.1        3.1.13-1     amd64        Microsoft .NET Core Runtime - 3.1.13 Microsoft.NETCore.App 3.1.13
+ii  dotnet-runtime-deps-3.1   3.1.13-1     amd64        dotnet-runtime-deps-3.1 3.1.13
+ii  dotnet-sdk-3.1            3.1.407-1    amd64        Microsoft .NET Core SDK 3.1.407
+ii  dotnet-targeting-pack-3.1 3.1.0-1      amd64        Microsoft.NETCore.App.Ref 3.1.0
+
+qsharp-runtime$ dpkg -l *aspnet*
```
Comment (swernli) [inline README.md:82]: Now that you confirmed the aspnetcore package isn't required, this part can be removed, along with the mention of aspnetcore-*-5.0 above.

**Review by swernli — APPROVED** [review]
One last removal, and I think these updates are good to go. Thanks for taking the time to do this!

File: README.md (hunk 22 above)
Comment (kuzminrobin) [reply README.md]: Done. Thanks.

File: src/QirRuntime/build.py [hunk 23]
```diff
@@ -1,102 +0,0 @@
-# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (DmitryVasilevsky) [inline src/QirRuntime/build.py:1]: Why are these files removed?

Comment (kuzminrobin) [reply src/QirRuntime/build.py:1]: The .py files are not applicable any more.
The .ps1 files are used for building and testing.
